### ISSUE TITLE
Planetwars planet selection confusion/changes

### ISSUE THREAD

**Issue by GoogleFrog** [ZeroK-RTS/Zero-K-Infrastructure#1855]
> Matchmaker picks 6 best targets. It makes sure one of them is always 2v2 planet (if such an option exists outside the 6 best targets).
> If there are no attackable planets at all, then 6 best non-attackable planets are selected. Note that this rule exists to prevent a total deadlock of PW, which is trivially possible for example using treaties. With this rule at least ticks progress and people can fix the issue.

I am still a bit unclear on how planet selection works, but I think changes are required. I will say how I think selection should work. Here are the two things that might be changes:
- Factions should be able to attack any valid target (no mechanically restrictive matchmaker).
- Factions should never be able to attack any invalid targets (no naive anti-deadlock).

There should be no matchmaker which restricts the set of planets that players are able to attack. If I recall correctly, in the previous round the matchmaker was not strictly restricting invasion options. Instead it was ranking the valid targets with some metric and picked the top six. Players could attack any valid planet via the web UI so, in effect, the top six list was just a method of organizing players. If the matchmaker UI is restricting options then a change is required.

Factions should not be able to attack anywhere simply by blocking of their normal attack options. This is surely something that reasonable factions would abuse, leading to ridiculous strategies. For example if faction A is dominating then faction B may ally and transfer planets to faction C so that faction B can randomly attack the infrastructure of A. It seems like every game with good FFA play on each side would have this tactic occur at least once (faction A might be just about to win). With the current mechanics faction B could just disable its own wormholes.

The question is now what to do about deadlocks. What we do relies on the answer to a few questions:
- Should treaties be unbreakable (I have been assuming that they are unbreakable).
- Should it be possible for a faction to own no planets?
- Should it be possible to disable wormholes (thus severing adjacency temporarily).
- Does an invasion have to occur for the turn to increment?

Here is my proposal:
 - Increment the turn upon the end of forfeit turns. That is, if it is faction A's turn to attack and they wait until their time runs out then do an economic tick (then it is faction B's turn to attack, as usual). The economic tick does everything as normal except that nobody receives any battle metal income. All factions still receive planet-based metal income, their structures produce ships, come closer to being online and, importantly, treaty turn counters are increased.
- If a faction has no valid targets at the beginning of their turn then make their attack timer drastically shorter. This should be a constant defined in the constants file. Try making it two minutes.

I think this proposal works because there are very few cases where a faction may want to forfeit their turn. Attacking gives them a chance at gaining IP and disabling structures. Attacking also gives them and another faction 100 metal but since this is a FFA they should be fine with both themselves and one other opponent gaining 100 metal. The only time accidental forfeiting will occur is in afk-PW but I don't think it turns too much, and besides, designing for afk may have problems.

A bonus of this proposal is that it makes activation times more reliable in afk-PW. Previously if two factions were afk you could effectively invade on every single turn. This made disabling defenses more powerful so further hurt afk factions.

**Comment by Licho1** [comment]
You are correct, matcher offers 6 but people can pick more options on website.

Forfeiting does increase a turn timer.

I can imagine all sorts of abuse however giving all planets from B to C is not an options.. C would never give them back and your faction dies.. how does this help ?

**Comment by Anarchid** [comment]
You get to capture planets from A and sell their infrastracture to buy yourselves a superweapon. 
Then donate those planets to C. This way you prevent A from reaching their imminent victory while benefiting both B and C. Perhaps B also gives you materiel like bombers and dropships so you can do your terror strikes more efficiently.

This is not conceptually different from how the unbreakable 100-turn Dynasty/Cyber peace pact was abused by Dynasty and Machines in the five-way PW.

**Comment by GoogleFrog** [comment]
> Forfeiting does increase a turn timer.

I thought that forfeiting on defense increases turn timer but on attack it did not. Forfeiting a defense is effectively not showing up to a battle, the participants still get 100 metal and IP changes hands. Are you sure that forfeiting on attack (as in, it is your turn to attack and you do nothing) increments the turn?

> I can imagine all sorts of abuse however giving all planets from B to C is not an options.. C would never give them back and your faction dies.. how does this help ?

B does not need to transfer all their planets to C. Instead, B allies with C in such a way that they cannot attack Cs planets. Then B transfers their outermost planets to C so that they are entirely surrounded and thus have no valid attack targets.

**Comment by Licho1** [comment]
forfeiting defense only increases timer, not attack.

**Comment by GoogleFrog** [comment]
Forfeit on attack should increase timer, then no deadlocks can occur.
